### PR TITLE
#gh10618 added C_BP_Relation endpoints

### PR DESCRIPTION
**Pull request by adi-stefan**
https://github.com/metasfresh/metasfresh/issues/10618

### CODE REVIEW

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	private final IBPRelationDAO bpRelationDAO = Services.get(IBPRelationDAO.class);
+
+	public BPRelationsService(@NonNull final JsonServiceFactory jsonServiceFactory)
+	{
+		this.jsonServiceFactory = jsonServiceFactory;
+	}
+
+	public JsonResponseBPRelationComposite getRelationsForPartner(@NonNull final OrgId orgId, @NonNull final IdentifierString bpartnerIdentifier)
+	{
+		final Optional<BPartnerComposite> bPartnerComposite = jsonServiceFactory.createRetriever().getBPartnerComposite(orgId, bpartnerIdentifier);
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:72]: Note: wondering if i makes sende to load the whole composite just for it's bpartner_id. on the other hand, they are cached, so i guess it's OK

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+	private final JsonServiceFactory jsonServiceFactory;
+	private final IBPRelationDAO bpRelationDAO = Services.get(IBPRelationDAO.class);
+
+	public BPRelationsService(@NonNull final JsonServiceFactory jsonServiceFactory)
+	{
+		this.jsonServiceFactory = jsonServiceFactory;
+	}
+
+	public JsonResponseBPRelationComposite getRelationsForPartner(@NonNull final OrgId orgId, @NonNull final IdentifierString bpartnerIdentifier)
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:70]: technically org is not requiered if the idneitier is a metasfreshId, but i think it's fine to always ask for the org

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		return builder.build();
+	}
+
+	public void createOrUpdateRelations(@NonNull final OrgId orgId,
+			@NonNull final IdentifierString bpartnerIdentifier,
+			@Nullable final IdentifierString locationIdentifier,
+			@NonNull final List<JsonRequestBPRelationTarget> relatesTo)
+	{
+		final Optional<BPartnerComposite> bPartnerComposite = jsonServiceFactory.createRetriever().getBPartnerComposite(orgId, bpartnerIdentifier);
+		final BPartnerComposite bpartner = Check.assumePresent(bPartnerComposite, "No bpartner found for identifier: " + bpartnerIdentifier);
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:127]: please throw a MissingResourceException

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				throw new AdempiereException("Unexpected type=" + locationIdentifier.getType());
+		}
+	}
+
+	private BPRelation fromJson(final BPartnerComposite sourceBPartner, final Optional<BPartnerLocation> location, final JsonRequestBPRelationTarget relatedBp)
+	{
+		final BPartnerId bpartnerId = sourceBPartner.getBpartner().getId();
+		final IdentifierString targetBpIdentifier = IdentifierString.of(relatedBp.getTargetBpartnerIdentifier());
+		final BPartnerComposite targetBpartnerComposite =
+				Check.assumePresent(jsonServiceFactory.createRetriever().getBPartnerComposite(sourceBPartner.getOrgId(), targetBpIdentifier)
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:169]: please throw a MissingResourceException instead

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+						, "No bpartner found for identifier: " + targetBpIdentifier);
+
+		final BPartnerId targetBPartnerId = targetBpartnerComposite.getBpartner().getId();
+		final IdentifierString targetLocationIdentifier = IdentifierString.ofOrNull(relatedBp.getTargetLocationIdentifier());
+		final Optional<BPartnerLocation> targetLocationOpt = getBpartnerLocation(targetBpIdentifier, targetLocationIdentifier, targetBpartnerComposite);
+		final BPartnerLocation targetLocation = Check.assumeNotNull(targetLocationOpt.orElseGet(getTargetPartnerLocationForRelation(targetBpartnerComposite, relatedBp.getShipTo(), relatedBp.getBillTo())), "Cannot identify location for target BPartner: {}, LocationIdentifier: {}", targetBpIdentifier, targetLocationIdentifier);
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:175]: please throw a MissingPropertyException instead

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+			{
+				location = bpartnerComposite.extractLocation(l -> l.getLocationType().getIsShipToOr(false));
+			}
+		}
+		if (!location.isPresent())
+		{
+			location = bpartnerComposite.extractLocation(Objects::nonNull);
+		}
+		final Optional<BPartnerLocation> result = location;
+		return () -> Check.assumePresent(result, "Cannot infer location for source BPartner");
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:226]: please throw a MissingPropertyException instead (rationale: it turned out that request is too unspecific as it lacks the explicit targetLocation which is needed for this particular bpartner)

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java (hunk 1 above)
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java:72]: Yeah, was thinking if a separate lower level cache only for bpartners, and logic to maintain it + lookup logic. Didn't make much sense.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java (hunk 3 above)
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java]: Changed to throw MissingResourceException

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java (hunk 4 above)
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java]: Changed to throw a MissingResourceException.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java (hunk 5 above)
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java]: Changed to throw a MissingPropertyException.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java (hunk 6 above)
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java]: Makes sense. I also noticed that the wording still referenced the source partner location, while the logic is applied only to the target bpartner. Updated to throw MissingPropertyException  and updated wording.

**Review by metas-ts — APPROVED** [review]

**Pull request merged**